⚰️ 📽️ Remove literal representations

mberr opened the pull request:
This PR removes the explicit class for literal representations, which can be replaced by a non-trainable `Embedding` with an pretrained embedding initializer. This reduces code duplication, but also gives access to all the gimmicks from `Embedding`, e.g., dropout and normalizers.

<del>The PR also deprecates the `create_init_from_pretrained` for a class, which allows code duplication reduction with the `LabelBasedInitializer`.</del>

mberr committed:
replace LiteralRepresentation by Embedding

using the init from pre-trained initializer and trainable=False

mberr committed:
Make PretrainedInitializer class

mberr committed:
use PretrainedInitializer for LiteralModel

mberr committed:
remove literal representation

trigger ci

mberr commented on src/pykeen/nn/init.py:
```diff
@@ -153,17 +196,11 @@ def create_init_from_pretrained(pretrained: torch.FloatTensor) -> Initializer:
-        return pretrained
-
-    return init_from_pretrained
+    warnings.warn(message="Please directly use `PretrainedInitializer`.", category=DeprecationWarning)
```
we could also directly remove this, since it was not yet included in any release (afaik)

cthoyt replied:
Just remove it

mberr replied:
4ee9e01

mberr commented on src/pykeen/nn/init.py:
```diff
@@ -120,6 +121,48 @@ def init_quaternions(
+    def __call__(self, x: torch.Tensor) -> torch.Tensor:
+        """Initialize the tensor with the given tensor."""
+        if x.shape != self.tensor.shape:
+            raise ValueError(f"shape does not match: expected {self.tensor.shape} but got {x.shape}")
```
not for this PR, but we could consider exposing `self.tensor.shape`, and maybe allow providing `embedding_dim=None` if the initializer provides a shape.

cthoyt replied:
could you please elaborate?

mberr replied:
If we explicitly want to use an initializer from pre-trained embeddings, then we need to set `embedding_dim` manually to the matching dimension, e.g., we want to use label-based init from `bert-base-cased`, then we will get 768-dimensional vectors. Right now, we would either need to create the initializer beforehand and inspect the stored `tensor`'s shape, or would need to lookup the dimension of the `bert-base-cased` model to determine how to set the `embedding_dim`.

The suggestion above was to allow setting `embedding_dim=None` in a model, which will then use the initializer's shape. If the initializer does not have a fixed shape, an error would be raised.

cthoyt replied:
Oh that's a bit round-about. It might cause more pain than it's worth to implement that

cthoyt replied:
Either way, thanks for explaining. we can maybe come back to it later if there's a need for it

mberr committed:
fix imports in test

trigger ci

mberr committed:
remove deprecated initializer creation

mberr committed:
small fixes

mberr committed:
trigger ci

cthoyt committed:
Fix typing

Technically this is a type annotation and you can't inherit from it, but MyPy is smart enough to see that this class basically follows the interface so no problemo

PyKEEN-bot committed:
Trigger CI

cthoyt committed:
Update init.py

Trigger CI

mberr committed:
Merge branch 'master' into remove-literal-representations

mberr committed:
Merge branch 'master' into remove-literal-representations

PyKEEN-bot committed:
Trigger CI
